cudax/include/cuda/experimental/__container/uninitialized_async_buffer.h

```diff
@@ -0,0 +1,213 @@
+  uninitialized_async_buffer(
+    _CUDA_VMR::async_resource_ref<_Properties...> __mr, const ::cuda::stream_ref __stream, const size_t __count)
+      : __mr_(__mr)
+      , __stream_(__stream)
+      , __count_(__count)
+      , __buf_(__count_ == 0 ? nullptr : __mr_.allocate_async(__get_allocation_size(__count_), __stream_))
+  {}
+
+  uninitialized_async_buffer(const uninitialized_async_buffer&)            = delete;
+  uninitialized_async_buffer& operator=(const uninitialized_async_buffer&) = delete;
+
+  //! @brief Move construction
+  //! @param __other Another \c uninitialized_async_buffer
+  uninitialized_async_buffer(uninitialized_async_buffer&& __other) noexcept
+      : __mr_(__other.__mr_)
+      , __stream_(_CUDA_VSTD::exchange(__other.__stream_, {}))
+      , __count_(_CUDA_VSTD::exchange(__other.__count_, 0))
+      , __buf_(_CUDA_VSTD::exchange(__other.__buf_, nullptr))
+  {}
+
+  //! @brief Move assignment
+  //! @param __other Another \c uninitialized_async_buffer
+  uninitialized_async_buffer& operator=(uninitialized_async_buffer&& __other) noexcept
+  {
+    if (__buf_)
+    {
+      __mr_.deallocate_async(__buf_, __get_allocation_size(__count_), __stream_);
+    }
+    __mr_     = __other.__mr_;
+    __stream_ = _CUDA_VSTD::exchange(__other.__stream_, {});
+    __count_  = _CUDA_VSTD::exchange(__other.__count_, 0);
+    __buf_    = _CUDA_VSTD::exchange(__other.__buf_, nullptr);
+    return *this;
+  }
+
+  //! @brief Destroys an \c uninitialized_async_buffer deallocating the buffer
+  //! @warning The destructor does not destroy any objects that may or may not reside within the buffer. It is the users
+  //! responsibility to ensure that all objects within the buffer have been properly destroyed.
+  ~uninitialized_async_buffer()
+  {
+    if (__buf_)
+    {
+      __mr_.deallocate_async(__buf_, __get_allocation_size(__count_), __stream_);
+    }
+  }
+
+  //! @brief Returns an aligned pointer to the buffer
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr pointer begin() const noexcept
+  {
+    return __get_data();
+  }
+
+  //! @brief Returns an aligned pointer to end of the buffer
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr pointer end() const noexcept
+  {
+    return __get_data() + __count_;
+  }
+
+  //! @brief Returns an aligned pointer to the buffer
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr pointer data() const noexcept
+  {
+    return __get_data();
+  }
+
+  //! @brief Returns the size of the buffer
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr size_t size() const noexcept
+  {
+    return __count_;
+  }
+
+  //! @brief Returns the \c async_resource_ref used to allocate the buffer
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE _CUDA_VMR::async_resource_ref<_Properties...> resource() const noexcept
+  {
+    return __mr_;
+  }
+
+  //! @brief Returns the stream used to allocate
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr ::cuda::stream_ref get_stream() const noexcept
```

Comment:
This is probably for a future PR that would add a way to re-bind a buffer to another stream, but maybe call this `get_bound_stream`?

Reply:
Yeah, lets move that to later